text/0004-new-pre-coordicide-consensus/0004-new-pre-coordicide-consensus.tex.md

```diff
@@ -0,0 +1,191 @@
++ Feature name: New Pre-Coordicide Consensus
++ Start date: 2020-02-10
++ RFC PR: [iotaledger/protocol-rfcs#0004](https://github.com/iotaledger/protocol-rfcs/pull/4)
+
+
+# Summary
+
+A change to the coordinator based protocol that should enable higher throughput of TPS and CTPS. It is achieved by doing the following changes:
+
+1. **White Flag** - Allow milestones to confirm conflicting bundles. Mutate the ledger state by deterministically ordering the bundles and ignoring conflicts.
+2. **New Node Tip Selection** - Nodes doing an almost random tip-selection will guarantee high throughputs and that honest txs will not likely be left behind. By adhering to certain constraints we will ensure that new transactions will add to the cumulative weight of existing transactions. A user will use `getTransactionsToApprove` API call with no parameters to invoke this tip selection.
+3. **New Compass Tip Selection** - Compass will use a heuristic to confirm a subtangle that has passed a certain threshold of cumulative weight (WIP).
+
+# Motivation
+
+The main motivations:
+1. **Defend against censorship attacks** - Conflicts will no longer block bundles from being approved.
+2.  **Make reattachments unneccessary** - As long as the network is not saturated, theoretically all bundles should be approved. And no bundle will be left behind.
+3. **Increase TPS** - Due to easy node tipselection the network throughput should increase.
+4. **Increase CTPS** - Due to the above, increase in TPS and no left-behinds, we expect CTPS to increase as well. 
+
+
+# Detailed design
+
+## White-Flag
+
+Let's define a `conflicting bundle` as a bundle that leads to a negative balance on an address if applied to the current ledger state.
+
+When a milestone is accepted by a node then the following algorithm to mutate the ledger state is performed: 
+
+1. Look for milestone whose index is the `current latest solid milestone index + 1`. Milestones must be applied in order!
+1. When the milestone is found start traversing down the tangle in [DFS](https://en.wikipedia.org/wiki/Depth-first_search) going first to the trunk and only then the branch. Stop at the transactions that has only approved parents.
+2. Start applying bundles to the diffmap only as you go up the recursion stack (when you climb up back from the leafs).
+3. Every time you attempt to apply a bundle to the diffmap:
+   - If it is invalid mark it as `seen` and `ignored` by the milestone. 
```

Comment:
What's the exact purpose of `seen` ? Doesn't `ignored`/`approved` imply `seen` ?